pkg/vault/contracts/Vault.sol

```diff
@@ -382,12 +389,9 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
 
         // (1) and (2): get raw amounts and check limits.
         if (vaultSwapParams.kind == SwapKind.EXACT_IN) {
-            // Swap fee is always a percentage of the amountCalculated. On ExactIn, subtract it from the calculated
-            // amountOut. Round up to avoid losses during precision loss.
-            locals.swapFeeAmountScaled18 = amountCalculatedScaled18.mulUp(swapState.swapFeePercentage);
-
-            // Need to update `amountCalculatedScaled18` for the onAfterSwap hook.
-            amountCalculatedScaled18 -= locals.swapFeeAmountScaled18;
+            // Restore the original input value; this function should not mutate memory inputs.
+            // At this point swap fee amounts have already been computed for EXACT_IN.
+            poolSwapParams.amountGivenScaled18 = swapState.amountGivenScaled18;
```

Comment:
> Since we've deleted other stuff from SwapLocals and freed up some space, consider not re-using amountGivenScaled18, so that we don't need to restore it? Could have amountGivenScaled18ForCalc/Mutable/PoolMath; something like that.

The problem is not about the locals, but about `poolSwapParams`. We need to send the whole struct to `onSwap`, and the only thing we're modifying is the `amountGivenScaled18`. So we either do something like this PR, or we copy the whole structure with a modified `amountGivenScaled18`.